control/grpc.py

```diff
@@ -236,41 +265,177 @@ def __init__(self, config: GatewayConfig, gateway_state: GatewayStateHandler, rp
         self.subsys_max_ns = {}
         self.host_info = SubsystemHostAuth()
 
-    def create_host_psk_file(self, subsysnqn : str, hostnqn : str, psk_value : str) -> str:
+    def get_directories_for_key_file(self, key_type : str, subsysnqn : str, create_dir : bool = False) -> []:
+        tmp_files = []
+        try:
+            tmp_files = os.listdir("/tmp")
+        except Exception:
+            self.logger.exception(f"Error listing files in /tmp")
+            return None
+
+        tmp_dirs = []
+        dir_prefix = f"{key_type}_{subsysnqn}_"
+        for tmp_file in tmp_files:
+            full_tmp_file_path = "/tmp/" + tmp_file
+            if not os.path.isdir(full_tmp_file_path):
+                continue
+            if not tmp_file.startswith(dir_prefix):
+                continue
+            tmp_dirs.insert(0, full_tmp_file_path)
+
+        if tmp_dirs:
+            return tmp_dirs
+
+        if not create_dir:
+            return None
+
+        tmp_dir_name = None
+        try:
+            tmp_dir_name = tempfile.mkdtemp(prefix=dir_prefix, dir="/tmp")
+        except Exception:
+            self.logger.exception("Error creating directory for key file")
+            return None
+        return [tmp_dir_name]
+
+    def create_host_key_file(self, key_type : str, subsysnqn : str, hostnqn : str, key_value : str) -> str:
         assert subsysnqn, "Subsystem NQN can't be empty"
         assert hostnqn, "Host NQN can't be empty"
-        assert psk_value, "PSK value can't be empty"
+        assert key_type, "Key type can't be empty"
+        assert key_value, "Key value can't be empty"
+
+        tmp_dir_names = self.get_directories_for_key_file(key_type, subsysnqn, create_dir = True)
+        if not tmp_dir_names:
+            return None
 
-        psk_dir = f"/tmp/psk/{subsysnqn}"
-        psk_file = f"{psk_dir}/{hostnqn}"
+        file = None
+        filepath = None
+        keyfile_prefix = f"{hostnqn}_"
         try:
-            os.makedirs(psk_dir, 0o755, True)
+            (file_fd, filepath) = tempfile.mkstemp(prefix=keyfile_prefix, dir=tmp_dir_names[0], text=True)
         except Exception:
-            self.logger.exception(f"Error creating directory {psk_dir}")
+            self.logger.exception("Error creating key file")
+            return None
+        if not filepath:
+            self.loger.error("Error creating key file")
             return None
         try:
-            with open(psk_file, 'wt') as f:
-                print(psk_value, end="", file=f)
-            os.chmod(psk_file, 0o600)
+            f = open(file_fd, "wt")
+            f.write(key_value)
+            f.close()
```

Comment:
Using context managers in Python (`with`) is better than explicitly calling for `close()`:
```suggestion
            with open(file_fd, "wt") as f:
                f.write(key_value)
```
For example it automatically deals with the case that an exception happens while writing. With the explicit `f.close()` an exception in `f.write()` would never reach that.

Reply:
Changed.